### ISSUE TITLE
files are copied twice and afterEachSync skips thousands of copy events

### ISSUE THREAD

**Issue by FossPrime** [hoperyy/sync-directory#27]
We are running fs-extra copySync on directories... which copies the whole directory and it's files, but then we proceed to re-copy the files. This is an issue when you have quotas like s3/gcp, metered cold storage and sandboxes.

https://codesandbox.io/s/sync-directory-n67ry?file=/test.mjs:2494-3223
```
async asyncSync () {
    const syncDir = (await import('./index.js')).async
    const dest = this.asyncDest
    let aec = 0
    let prevDirCount = 0
    let prevPath = ''
    await syncDir(this.src, dest, {
      type: 'copy',
      afterEachSync: async ({relativePath}) => {
        aec++
        const dirCount = Tests.getDirFCount(dest)
        if (prevDirCount+1 !== dirCount) {
          console.log(prevDirCount, dirCount, prevPath, relativePath)
        }
        prevDirCount = dirCount
        prevPath = relativePath
      }
    })
    this.asyncAfterEachCount = aec
    assert(this.srcSize, Tests.getDirKiB(this.asyncDest))
    assert(this.srcFCount, Tests.getDirFCount(this.asyncDest))
    logger('async passed.')
  }
```

```
[Tests] cli passed.
1024 3072 /999.bin /child
3072 3072 /child /child/0.bin
3072 3072 /child/0.bin /child/1.bin
3072 3072 /child/1.bin /child/10.bin
```

## Potential solution:
```
const copyDirFiles = (srcPath, targetPath) => {
  try {
    if (isDirectoryUtil(srcPath)) {
      fs.mkdirSync(targetPath)
    } else {
      fse.copySync(srcPath, targetPath)
    }
  } catch (err) {
    console.error({
      message: 'Error copying "' + srcPath + '" to "' + targetPath + '"',
      code: 1000
    })
    throw new Error(err)
  }
}
```
Yields the correct: (the file counter function explicitly subtracts directories)
```
[Tests] 2021-10-16T04:28:35.473Z
[Tests] cli passed.
1024 1024 /999.bin /child
2048 2048 /child/999.bin /child/grandchild
[Tests] async passed.
[Tests] sync passed.
[Tests] Runtime: 90655ms
[Tests] DEBUG:  src /tmp/sdsrc-Hlyc6A
[Tests] DEBUG:  cliDest /tmp/sdcli-Rsxvxu
[Tests] DEBUG:  syncDest /tmp/sdsync-L39QYn
[Tests] DEBUG:  asyncDest /tmp/sdasync-FMeeqh
[Tests] DEBUG:  srcSize 13404
[Tests] DEBUG:  srcFCount 3072
[Tests] DEBUG:  asyncAfterEachCount 3074
[Tests] DEBUG:  debug true
[Tests] DEBUG:  runtime 90655.3438129425
```

We can also add a callback overload to cheaply run async functions as callback hell functions

**Comment by hoperyy** [comment]
Yes.

Directory sync was ignored from `afterEachSync`. I'll fix it soon. @rayfoss

**Comment by rayfoss** [comment]
> Yes.
> 
> Directory sync was ignored from `afterEachSync`. I'll fix it soon. @rayfoss

It's working, but extremely inefficiently. Instead of performing 3072 write operations, it performs 3072+2048+1024 write operations. The test.mjs artifact is a folder that looks like this

- source
    - 1024 files
    - child
        - 1024 files
        - grandchild
            - 1024 files
 
Don't sweat over it, my usecase is more of an rsync situation with strict quotas, on exclusively Debian environments. As such ill probably need to borrow local-sync and customize it with typescript, quota tests, perf optimizations, skip-updated support and/or checksums.

**Comment by hoperyy** [comment]
Well I just published `sync-directory@4.0.0`.

@rayfoss 

More powerful `afterEachSync()` is supported: 

```js
afterEachSync({ eventType, nodeType, relativePath, srcPath, targetPath }) {
            // eventType: "init:hardlink" / "init:copy" / "add" / "change" / "unlink" / "unlinkDir"
            //          - init type: "init:hardlink" / "init:copy"
            //          - watch type: "add" / "change" / "unlink" / "unlinkDir"
            // nodeType: "file" / "dir"
            // relativePath: relative file/folder path
            // srcPath: absolute src file/folder path
            // targetPath: absolute target file/folder path
}
```
